File: ruby/src/coin-change-ii.rb
# @param {Integer} amount
# @param {Integer[]} coins
# @return {Integer}
def change(amount, coins)
  coins.sort! { _2 <=> _1 }
  cache = Array.new(coins.size) { Array.new(amount + 1) }
  recur(coins, cache, amount, 0)
end

def recur(coins, cache, amount, i)
  return 1 if amount.zero?
  return 0 if amount.negative?
  return 0 if i >= coins.size

  cache[i][amount] ||= (i...coins.size).sum do |next_i|
    recur(coins, cache, amount - coins[next_i], next_ix)
  end
end

File: python3/coin-change-ii.py
class Solution:
    """
    >>> Solution().change(5, [1, 2, 5])
    4
    """
    def change(self, amount: int, coins: list[int]) -> int:
        coins.sort()
        ways = [0] * (amount + 1)
        ways[0] = 1
        for coin in coins:
            for current_amount in range(amount):
                if current_amount + coin > amount:
                    break
                ways[current_amount + coin] += ways[current_amount]
        return ways[amount]
